### PR TITLE
Embedded slip condition

### PR DESCRIPTION
**Pull request by rubenzorrilla**
A Navier-slip boundary condition has been implemented to be used together with the embedded element.

### CODE REVIEW

File: applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h [hunk 1]
```diff
@@ -534,17 +434,22 @@ class EmbeddedNavierStokes : public NavierStokes<TDim, TNumNodes>
...
+    */
+    void ComputeSplitting(const ElementDataType& rData,
+                          const array_1d<double,TNumNodes>& rDistances,
+                          ElementSplittingDataStruct& rSplittingData)
     {
 
-        unsigned int ndivisions = 1;
-        VectorType NodalDistances = distances;
-        MatrixType NoSplitGradients = data.DN_DX;
+        rSplittingData.edge_areas = ZeroVector((TDim-1)*3);         // Should be initialized to the edge number
```
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h:449]: those two ways do not reallocate memory

    edge_areas.clear() 
    noalias(edge_areas) = ZeroVector(...)

File: applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h [hunk 2]
```diff
@@ -658,240 +552,201 @@ class EmbeddedNavierStokes : public NavierStokes<TDim, TNumNodes>
...
-
-        for (unsigned int icut = 0; icut<cut_points; icut++)
+        for (unsigned int icut = 0; icut<rSplittingData.ncutpoints; icut++)
         {
-            double weight = cut_areas(icut);
-            intersection_area += weight;
-
-            aux_cut = row(Ncontainer_cut, icut);
+            const double weight = rSplittingData.cut_edge_areas(icut);
+            const VectorType aux_cut = row(rSplittingData.N_container_cut, icut);
```
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h]: i believe you know the size at compile type of aux_cut. if so try if possible to use a stack array instead

File: applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h [hunk 3]
```diff
@@ -1099,162 +915,757 @@ class EmbeddedNavierStokes : public NavierStokes<TDim, TNumNodes>
...
+        GetPreviousSolutionVector(rData, prev_sol);
 
-        // Add all the boundary intersection terms
-        if (ndivisions > 1)
+        // Nitsche coefficient computation
+        const double eff_mu = BaseType::ComputeEffectiveViscosity(rData);
+        // const double cons_coef = 10*eff_mu/rData.h;
+
+        // Compute the element average velocity norm
+        std::vector<double> avg_vel;
```
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h]: you sure you need an std::vector here?

File: applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h [hunk 4]
```diff
@@ -1099,162 +915,757 @@ class EmbeddedNavierStokes : public NavierStokes<TDim, TNumNodes>
...
-            constexpr unsigned int MatrixSize = TNumNodes*BlockSize;   // Matrix size
+            v_norm += avg_vel[i]*avg_vel[i];
+        }
+        v_norm = std::sqrt(v_norm);
+
+        // Compute the element average density
+        double avg_rho = 0.0;
+        for (unsigned int j=0; j<TNumNodes; ++j)
+        {
+            avg_rho += (this->GetGeometry())[j].FastGetSolutionStepValue(DENSITY);
```
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h]: have an rData.densities

**Review by RiccardoRossi — APPROVED** [review]
if you

File: applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h [hunk 5]
```diff
@@ -204,68 +228,13 @@ class EmbeddedNavierStokes : public NavierStokes<TDim, TNumNodes>
         }
```
Comment (jcotela) [inline applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h]: I'm not sure if this is needed (you are already resizing the matrices above)

File: applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h [hunk 6]
```diff
@@ -277,215 +246,146 @@ class EmbeddedNavierStokes : public NavierStokes<TDim, TNumNodes>
...
+        KRATOS_TRY;
 
         // Base element check
         int ErrorCode = NavierStokes<TDim, TNumNodes>::Check(rCurrentProcessInfo);
         if(ErrorCode != 0) return ErrorCode;
 
         // Specific embedded element check
         if(DISTANCE.Key() == 0)
-            KRATOS_THROW_ERROR(std::invalid_argument,"DISTANCE Key is 0. Check if the application was correctly registered.","");
+            KRATOS_ERROR << "DISTANCE Key is 0. Check if the application was correctly registered.";
```
Comment (jcotela) [inline applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h]: You can also check that DISTANCE is assigned to the nodal database (SolutionStepDataHas())

**Review by jcotela — APPROVED** [review]

File: applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h (hunk 2 above)
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h]: Done. Changed by and array_1.

File: applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h (hunk 3 above)
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h]: Actually not... Removed.

File: applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h (hunk 4 above)
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h]: Done!

File: applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h (hunk 5 above)
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h]: You're right. It is not needed.

File: applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h (hunk 6 above)
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h]: Added.

File: applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h (hunk 1 above)
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h:449]: @RiccardoRossi In fact, in here is where I allocate the memory. To use an array_1d would imply to modify the SplitTetrahedraUtilities() in the core, which is something that I think we do not want to do...

**Pull request merged**